### PR TITLE
NVME support

### PR DESCRIPTION
**Pull request by drake7707**
#### Description

See issue  #2916, by using pySMART instead of the old pySMART-smartx the nvme attributes can also be fetched

Someone is going to need to test if this impacts existing drives, it shouldn't but i don't have an hdd in my server.

image

#### Resume

* Bug fix: no
* New feature: yes
* Fixed tickets: #2916, #2539

### CODE REVIEW

**Comment by nicolargo** [comment]
Hi @drake7707 

Thanks for the PR !

Is anyone with a  Smart driver and a non NVME disk can test the PR ?

File: glances/plugins/smart/__init__.py [hunk 1]
```diff
@@ -41,6 +41,7 @@
 # Import plugin specific dependency
 try:
     from pySMART import DeviceList
+    from pySMART.interface.nvme import NvmeAttributes
```
Comment (nicolargo) [inline glances/plugins/smart/__init__.py]: Change the comment at the top of the file as pySMART.smartx is no longer used.

**Comment by drakarah** [comment]
> Hi @drake7707
> 
> Thanks for the PR !
> 
> Is anyone with a Smart driver and a non NVME disk can test the PR ?

I managed to find an USB adapter that supports SMART data for a SATA drive and hooked it up to my server, mapped it through to the docker glances container and i'm getting:


image
image

Which I assume is correct?

**Comment by webdjoe** [comment]
Thank you for the PR! I need to debug some more, but I'm getting the following error with a SATA drive: 

```
Traceback (most recent call last):
  File "/prodrive/apps/glances/.venv/bin/glances", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/prodrive/apps/glances/.venv/lib/python3.12/site-packages/glances/__init__.py", line 191, in main
    start(config=core.get_config(), args=core.get_args())
  File "/prodrive/apps/glances/.venv/lib/python3.12/site-packages/glances/__init__.py", line 150, in start
    mode = GlancesMode(config=config, args=args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/prodrive/apps/glances/.venv/lib/python3.12/site-packages/glances/webserver.py", line 32, in __init__
    self.stats.update()
  File "/prodrive/apps/glances/.venv/lib/python3.12/site-packages/glances/stats.py", line 287, in update
    self.update_plugin(p)
  File "/prodrive/apps/glances/.venv/lib/python3.12/site-packages/glances/globals.py", line 564, in inner
    return _func(weakref.ref(self), *args, ttl_hash=_get_ttl_hash(ttl), **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/prodrive/apps/glances/.venv/lib/python3.12/site-packages/glances/globals.py", line 560, in _func
    return func(_self(), *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/prodrive/apps/glances/.venv/lib/python3.12/site-packages/glances/stats.py", line 274, in update_plugin
    self._plugins[p].update()
  File "/prodrive/apps/glances/.venv/lib/python3.12/site-packages/glances/plugins/plugin/model.py", line 1151, in wrapper
    ret = fct(self, *args, **kw)
          ^^^^^^^^^^^^^^^^^^^^^^
  File "/prodrive/apps/glances/.venv/lib/python3.12/site-packages/glances/plugins/plugin/model.py", line 1172, in wrapper
    logger.debug(f"{class_name} {class_module} {fct.__name__} return {ret} in {duration} seconds")
                                                                     ^^^^^
  File "/prodrive/apps/glances/.venv/lib/python3.12/site-packages/pySMART/interface/nvme/__init__.py", line 499, in __repr__
    return ("{0:>2} {1:18}{2:29}{3:16}{4:>13}{5:>5}{6:>4}{7:>4}{8:>4}".format(

```

**Comment by drakarah** [comment]
@webdjoe That might be a pySMART bug, from what I can tell it's trying to serialize the NVMESelfTest to a string which calls `__repr__` and the formatting fails. It shouldn't even attempt to parse the smartctl output of a SATA drive as an NVME.

Can you share your smartctl output?

**Comment by webdjoe** [comment]
I did some more debugging and found that the issue is that some self-test object attributes are `None`, which causes the `__repr__` method to fail. These are the attributes of the pySMART  `nvmeSelfTest` object that cause it to fail here https://github.com/truenas/py-SMART/blob/935438422683eff5a0ff7103f6d0802077b5b94d/pySMART/interface/nvme/__init__.py#L499:

```
<pySMART.interface.nvme.NvmeSelfTest object at 0x7e3a86af10a0>
code = '-'
description = 'Short'
failingLBA = None
nsid = None
num = 0
powerOnHours = 9136
progress = 100
sct = '-'
seg = None
status = 'Completed without error'
```

This could be addressed in the pySMART library, but maybe the self test attribute can be excluded in Glances since I do not believe it is used.


If needed, the smartctl output is: 

```
=== START OF INFORMATION SECTION ===
Model Number:                       CT500P3PSSD8
Serial Number:                      2313E6C0A903
Firmware Version:                   P9CR40A
PCI Vendor/Subsystem ID:            0xc0a9
IEEE OUI Identifier:                0x00a075
Controller ID:                      1
NVMe Version:                       1.4
Number of Namespaces:               1
Namespace 1 Size/Capacity:          500,107,862,016 [500 GB]
Namespace 1 Formatted LBA Size:     512
Namespace 1 IEEE EUI-64:            6479a7 7780000032
Local Time is:                      Mon Dec 15 02:39:57 2025 UTC
Firmware Updates (0x12):            1 Slot, no Reset required
Optional Admin Commands (0x0017):   Security Format Frmw_DL Self_Test
Optional NVM Commands (0x005e):     Wr_Unc DS_Mngmt Wr_Zero Sav/Sel_Feat Timestmp
Log Page Attributes (0x06):         Cmd_Eff_Lg Ext_Get_Lg
Maximum Data Transfer Size:         64 Pages
Warning  Comp. Temp. Threshold:     85 Celsius
Critical Comp. Temp. Threshold:     95 Celsius

Supported Power States
St Op     Max   Active     Idle   RL RT WL WT  Ent_Lat  Ex_Lat
 0 +     6.00W  0.0000W       -    0  0  0  0        0       0
 1 +     3.00W  0.0000W       -    0  0  0  0        0       0
 2 +     1.50W  0.0000W       -    0  0  0  0        0       0
 3 -   0.0250W  0.0000W       -    3  3  3  3     5000    1900
 4 -   0.0030W       -        -    4  4  4  4    13000  100000

Supported LBA Sizes (NSID 0x1)
Id Fmt  Data  Metadt  Rel_Perf
 0 +     512       0         1
 1 -    4096       0         0

=== START OF SMART DATA SECTION ===
SMART overall-health self-assessment test result: PASSED

SMART/Health Information (NVMe Log 0x02)
Critical Warning:                   0x00
Temperature:                        36 Celsius
Available Spare:                    100%
Available Spare Threshold:          5%
Percentage Used:                    4%
Data Units Read:                    2,673,106 [1.36 TB]
Data Units Written:                 7,937,125 [4.06 TB]
Host Read Commands:                 14,243,612
Host Write Commands:                239,481,030
Controller Busy Time:               200
Power Cycles:                       40
Power On Hours:                     20,791
Unsafe Shutdowns:                   29
Media and Data Integrity Errors:    0
Error Information Log Entries:      186
Warning  Comp. Temperature Time:    0
Critical Comp. Temperature Time:    0
Temperature Sensor 1:               36 Celsius
Temperature Sensor 2:               41 Celsius
Temperature Sensor 8:               36 Celsius

Error Information (NVMe Log 0x01, 16 of 16 entries)
Num   ErrCount  SQId   CmdId  Status  PELoc          LBA  NSID    VS  Message
  0        186     0  0x6004  0x4005  0x028            0     0     -  Invalid Field in Command

Self-test Log (NVMe Log 0x06)
Self-test status: No self-test in progress
Num  Test_Description  Status                       Power_on_Hours  Failing_LBA  NSID Seg SCT Code
 0   Short             Completed without error                9136            -     -   -   -    -
 1   Extended          Aborted: Self-test command             9136            -     -   -   -    -

```

**Comment by drakarah** [comment]
Ill add a try catch around the serialisation of each attribute so if for some reason the attribute ends up being an object that can't be serialised to string for whatever reason it should just be omitted instead of crashing the whole stack.

**Comment by drakarah** [comment]
@webdjoe Can you try again with the latest commit?

**Comment by webdjoe** [comment]
Sorry for the delay. You are on the right track, but I think you are trying to do something like this in https://github.com/nicolargo/glances/blob/9d4bc76304a4c205c55c47b1a99ab77155fdf51b/glances/plugins/smart/__init__.py#L132


```python

        if isinstance(dev.if_attributes, NvmeAttributes):
            idx = 0
            for attr in dev.if_attributes.__dict__.keys():
                try:
                    attrib_dict = convert_nvme_attribute_to_dict(attr,  dev.if_attributes.__dict__[attr])
                    if dev.if_attributes.__dict__[attr] is not None:
                        serialized = str(dev.if_attributes.__dict__[attr])
                        idx +=1
                        stats[-1][idx] = attrib_dict
                except Exception as e:
                    logger.debug(f'Unable to serialize attribute {attr} from NVME')


```

this works perfectly. It also prevents empty attributes from being added.

@nicolargo  Are you ok with stripping leading and trailing underscores for attribute names?  Like `_temperature` to `temperature`

**Comment by drakarah** [comment]
@webdjoe If you only increment idx for non empty attributes then whenever an attribute becomes empty/nonempty the entire index shifts, and any integration (such as home assistant, https://github.com/nicolargo/glances/issues/2916#issuecomment-3401485140) that took the JSON path with the index to get a specific attribute value will fail. 

I'm not really sure why it used indices as keys in the first place but the SATA code did so I wanted to make it the same

**Comment by webdjoe** [comment]
Then to maintain consistency, this would be the edit:

```python
        if isinstance(dev.if_attributes, NvmeAttributes):
            idx = 0
            for attr in dev.if_attributes.__dict__.keys():
                idx +=1
                attrib_dict = convert_nvme_attribute_to_dict(attr,  dev.if_attributes.__dict__[attr])
                try:
                    if dev.if_attributes.__dict__[attr] is not None:
                        serialized = str(dev.if_attributes.__dict__[attr])
                except Exception as e:
                    logger.debug(f'Unable to serialize attribute {attr} from NVME')
                    attrib_dict['value'] = None
                    attrib_dict['raw'] = None
                finally:

                    stats[-1][idx] = attrib_dict

```

**Comment by drakarah** [comment]
I've updated the code with your code snippet and merged the latest changes from develop.

It's unfortunate that NvmeAttributes does not expose the raw smartctl key-value pairs parsed but instead uses properties. I suppose making a reverse lookup of https://github.com/truenas/py-SMART/blob/935438422683eff5a0ff7103f6d0802077b5b94d/pySMART/interface/nvme/__init__.py#L525 would make the labels a bit neater

**Comment by drakarah** [comment]
In fact let me just make a lookup map

image

**Comment by nicolargo** [comment]
Hi @drake7707 

Thanks for your work on this feature.

Some improvements:

1) Make the list of attributes to display configurable from the Glances configuration file (see what is done for the disable_stats of the processlist plugin):

Non relevant example:

    [smart]
    hide_attributes=hostReadCommands,hostWriteCommands,warningTemperatureTime,criticalTemperatureTime

or

    [smart]
    show_attributes=dataUnitsRead,availableSpare

2) Use auto_unit (globals.py) function to display numerical value in "human reading"  

    Bytes reads:                        1400000000

will be displayed as:

    Bytes reads:                                      1.4G

3) Update documentation: https://github.com/nicolargo/glances/blob/develop/docs/aoa/smart.rst

**Comment by drakarah** [comment]
@nicolargo I've made the changes you requested

The hide_attributes is based on name and not the dictionary key (basically the labels that are shown should be used to hide them).

For the auto_unit i've only used it for ["bytesWritten", "bytesRead", "dataUnitsRead", "dataUnitsWritten", "hostReadCommands", "hostWriteCommands" ]. Any other reported stat won't be auto_unit'd.

The auto_unit is only applied to the curses output, like the other plugins do. I don't see where the web interface does that?

**Comment by webdjoe** [comment]
Hey @nicolargo is there anything I can do to help move this forward? Thank you for all the hard work on this amazing library.

**Comment by nicolargo** [comment]
@drake7707 / @webdjoe 

Just tested the latest version of the PR on my side.

1) In the TUI, it displays:

image

In the WEBUI:

image

It's not a big deal but i will prefer to have both view with the same look and feel.

2) The documentation (./docs/aoa/smart.rst) should be updated to warn the user that this plugin only work with sudo/root account.

Let me know if you can do it.  I can merge without those changes, no problem.

**Comment by drakarah** [comment]
How do i change the web UI without affecting the api format? When I looked for references on auto_unit I saw that plugins only used it in the curses output. Can you point me to where I need to change the formatting?

**Comment by nicolargo** [comment]
@drake7707 you can have a look to another plugin (for example glances/outputs/static/js/components/plugin-fs.vue). You should use methods implemented in "$filters" (glances/outputs/static/js/filters.js).

**Comment by drakarah** [comment]
@nicolargo  ok thanks for the pointer, i figured out the vue part and modified it to format it the same way. I've commited the changes 

>      The documentation (./docs/aoa/smart.rst) should be updated to warn the user that this plugin only work with sudo/root account.

It already does?
>.. warning::
>    This plugin needs administrator rights. Please run Glances as root/admin.

**Comment by nicolargo** [comment]
Ok perfect the WebUI displays information like the TUI does :)

Before merging to develop, i have made some change on your code:
- Lint/format
- reduce cyclomatic complexity 
- correct some typo
- add your Github user to the Copyright

The PR is now merged into develop

Thanks for your contrib @drake7707 !
